Refactoring

EmanuelUnity3D opened the pull request:
I did a variety of things here.

- I brought code up to PEP8 standard and added documentation.
- I made some bug fixes where I saw the need for them.
- I removed type checking from the functions and added domain checks where those were missing.
- I restructured the parser, added documentation to it, and added error handling to it.
- I made it possible to insert equation elements at the position of the cursor instead of the end of the field.

That being said, there are some things that I did not do.

- I did not add testing for the parser.
- I did not document the test files or bring them up to PEP8 standard.
- I did not add the additional tests that we will need to have a complete testing suite.

In summary, testing is probably the last thing that we need to emphasize before we can call it a wrap on the Python code.

VisusAdAstra commented on parse.py:
```diff
@@ -0,0 +1,209 @@
+from functions import trigonometry
+import exceptions.exceptions as exceptions
+
+from pyparsing import *
```
maybe not a good idea as it will put irrelevant things into the code, also can cause name conflict

EmanuelUnity3D replied:
The alternative is the long list of imports that we had before. I could change it back if you guys are really worried about the sanity of this approach but we haven't had any conflicts yet and the Python code seems pretty complete.

cpotteck replied:
I would say revert this to be thorough

SwatiPareek commented on calculate.py:
```diff
@@ -1,22 +1,30 @@
+        expression = expression.replace('√', 'sqrt')
+        # Evaluate expression
+        return parser.evaluate(expression)
+    except Exception as e:
```
Do we want to catch specific exceptions later?

EmanuelUnity3D replied:
I have this here as a contingency. I don't even know if there are any circumstances under which an exception could reach here. If there are, the important part is that they are passed to the calculator somehow for us to see. The more specific handling is done in the parser at the moment.

SwatiPareek commented on functions/exponents_and_logs.py:
```diff
@@ -1,76 +1,209 @@
+
+    Args:
+        base (float): Base of power to be returned
+        exponent (float): Exponent of power to be returned
```
exponent(int)

SwatiPareek commented on functions/common.py:
```diff
@@ -1,26 +1,108 @@
     return result
 
 
-def mod(dividend: float, divisor: float) -> float:
```
Why was mod function removed, just curious.

EmanuelUnity3D replied:
I didn't see it being used and I figured we should use % instead anyways (unless the mod function does something different). Let me know if I should revert it though.

SwatiPareek commented on functions/common.py:
```diff
@@ -1,26 +1,108 @@
+    Returns:
+        bool: True if 'argument' is even and false otherwise
+    """
+    return not (argument & 1)
```
May want to do type checking here and raise exception when argument is not an int. 
This would also apply for is_odd

EmanuelUnity3D replied:
I think I left it out to avoid spending cycles on it but you're right it probably should be there. I'll add it.

SwatiPareek commented on functions/exponents_and_logs.py:
```diff
@@ -1,76 +1,209 @@
-        e += numerator / denominator
-    return e
 
+def exp(exponent: float = 1) -> float:
```
Function should not be named exp if its calculating e^x, this isn't intuitive and can be quite confusing

EmanuelUnity3D replied:
I thought that generate_e was misleading because it can generate any power of e (not necessarily just e itself). I thought of changing it to generate_power_of_e but then I decided to just follow the naming scheme used by math libraries like Python and Java's. They both use exp to refer to powers of e and pow to refer to powers in general. I figured that this was both intuitive and short enough to also be used as an identifier in the calculator interface.

SwatiPareek replied:
I agree the original was misleading but for those unfamiliar with the inner workings of the Python/Java math libs, exp doesn't clearly articulate we're using e. I would prefer to use generate_power_of_e and generate_e for e^x and e respectively. If we decide to create a separate file for pi/e and other constants, having the generate_%constant% convention would be more readable than using a distinct naming strategy for e. 

I don't consider this a major issue though. Lets run it by a few more people. If its nor confusing to the majority of us, I'm fine with leaving it as is. I just don't know how familiar the TAs may be of these conventions.

EmanuelUnity3D replied:
I could get with using generate_power_of_e for the sake of consistency. I had kind of started doing something similar with pow_pi and pow_10. Still though, we would need to choose a name for the calculator interface itself. I like exp because it's short and established by other libraries but if it isn't intuitive to the others then I'm probably alone on this. Maybe we could change it to pow_e or something?

cpotteck replied:
I like the prefix `pow` with specifics after

EmanuelUnity3D replied:
Sounds good, I will adopt this convention.

SwatiPareek commented on functions/exponents_and_logs.py:
```diff
@@ -1,76 +1,209 @@
+    fractional_part_of_exponent = exponent - integer_part_of_exponent
+    
+    # Approximate power of e.
+    e = exp_taylor(1)
```
Same comment of exp_taylor --> not necessarily intuitive that we're calculating e

EmanuelUnity3D replied:
Same comment as above though I want to hear your take on it.

cpotteck replied:
since this is the generic power of e, `pow_e` could fit here

SwatiPareek commented:
Thinking that maybe we should move functions involving generation of pi and e to common or even better a separate file for transcendental constants.

EmanuelUnity3D commented:
> plz add test suite to ensure proper functionality

Testing has been assigned to the team as a whole. I think it should be a separate branch. That being said, I don't see any problem with branching off of this branch if you guys want to.

SwatiPareek commented:
> 
> 
> > plz add test suite to ensure proper functionality
> 
> Testing has been assigned to the team as a whole. I think it should be a separate branch. That being said, I don't see any problem with branching off of this branch if you guys want to.

I think @VisusAdAstra was referring to using tests as a way to ensure we didn't break anything with this refactoring. 
The unit tests should cover individual functions, as for testing the application itself, would it be easier to just check into this branch and run the application as is to determine it behaves as expected? I didn't go through the parser logic thoroughly yet but if refactoring has changed some of that logic, we may want to add some basic tests for the parser in this PR if that's possible (This does not have to be just you as other people can contribute to this PR while its open)

EmanuelUnity3D commented:
> > > plz add test suite to ensure proper functionality
> > 
> > 
> > Testing has been assigned to the team as a whole. I think it should be a separate branch. That being said, I don't see any problem with branching off of this branch if you guys want to.
> 
> I think @VisusAdAstra was referring to using tests as a way to ensure we didn't break anything with this refactoring.
> The unit tests should cover individual functions, as for testing the application itself, would it be easier to just check into this branch and run the application as is to determine it behaves as expected? I didn't go through the parser logic thoroughly yet but if refactoring has changed some of that logic, we may want to add some basic tests for the parser in this PR if that's possible (This does not have to be just you as other people can contribute to this PR while its open)

I assumed that we would rewrite the tests to use the parser when we refactor the test suite. That way we can test both features at once. I don't know if the rest of the team is onboard though. I brought it up once during the last meeting but I don't remember the response.

SwatiPareek commented:
> ctor the test suite. That way we can test both features at once. I don't know if the rest of the team is onboard though. I brought it up once during the last meeting but I don't remember the response.

That stretches the definition of a unit test though. We may want to have some sort of integration/sanity test that would test the parser with the functions but this should not replace our existing unit tests. I believe they serve a different purpose.

EmanuelUnity3D commented:
> > ctor the test suite. That way we can test both features at once. I don't know if the rest of the team is onboard though. I brought it up once during the last meeting but I don't remember the response.
> 
> That stretches the definition of a unit test though. We may want to have some sort of integration/sanity test that would test the parser with the functions but this should not replace our existing unit tests. I believe they serve a different purpose.

Fair enough. I had considered the fact that this would make debugging failed tests a little more difficult but I figured that the larger test suite for the parser might be worth it. I think you have a good point though.

cpotteck commented:
> > > ctor the test suite. That way we can test both features at once. I don't know if the rest of the team is onboard though. I brought it up once during the last meeting but I don't remember the response.
> > 
> > 
> > That stretches the definition of a unit test though. We may want to have some sort of integration/sanity test that would test the parser with the functions but this should not replace our existing unit tests. I believe they serve a different purpose.
> 
> Fair enough. I had considered the fact that this would make debugging failed tests a little more difficult but I figured that the larger test suite for the parser might be worth it. I think you have a good point though.

Yeah I think we can keep maintaining tests at every layer, which is good practice (I think). The code coverage tool can help us with this.

cpotteck commented on functions/common.py:
```diff
@@ -1,26 +1,122 @@
+    return not (argument & 1)
+
+
+def is_odd(argument: int) -> bool:
```
just wondering why `is_odd` isn't just `not is_even` (or vice-versa)? isn't there some duplication here

EmanuelUnity3D replied:
To be honest, I don't remember the logic here. I think I implemented both just to be thorough. I probably figured that is_even is easier to read than not is_odd. Typically, I wouldn't implement either. Instead, I would just write the code in line.

cpotteck replied:
I meant more stuff like type checking is duplicated when its exactly the same in both cases, and that the body of `is_even` could just be `return not is_odd` or vice-versa. Don't have strong opinions here though so its non-blocking

cpotteck replied:
I meant more stuff like type checking is duplicated when its exactly the same in both cases, and that the body of `is_even` could just be `return not is_odd` or vice-versa. Don't have strong opinions here though so its non-blocking

SwatiPareek replied:
I don't type checking is enforced through definition params iirc. 
Since the def of is_odd will never change, redundancy isn't inherently problematic. 
It just boils down to preference.

EmanuelUnity3D replied:
I think he means the actual type checking being done in the function body. I kept the functions separate because they're pretty simple and I don't want to add extra functions calls. Especially on a small function like this. Truth be told, I could have done it I just prefer not to for the sake of performance (not that we're having any performance issues though).

cpotteck commented on tests/test_exponents_and_logs.py:
```diff
@@ -1,5 +1,5 @@
-import exceptions.exceptions as the_exception
+import exceptions.exceptions as exceptions
 import functions.exponents_and_logs as exp
 import functions.trignometry as trig
```
typo `trignometry` > `trigonometry`

EmanuelUnity3D commented:
> *Also need to rename `test_trignometry.py` to `test_trigonometry.py`. Make sure tests run 😜 .
> 
> Looks good overall after reading through code and doing some manual testing, added a few comments for questions/suggestions/changes.

Yeah I think I changed the name of that file at the end when I had moved on to working on the parser. I should have rerun the unit tests regardless though. My bad.

VisusAdAstra commented on calculate.py:
```diff
@@ -1,22 +1,30 @@
 import sys
 import re
-from nparser import Parser
+from parse import Parser
```
just a small detail but do you think parser will be more correct name?

EmanuelUnity3D replied:
Having the file name be the same as the class name makes importing complicated. I think that that is why the original file name had an n prefixed to it.

VisusAdAstra commented on parse.py:
```diff
@@ -0,0 +1,222 @@
+        is_negative = False
+        for e in tokens:
+            if e == '-':
+                is_negative = not is_negative
```
is_negative =True isn't correct here???

EmanuelUnity3D replied:
I don't understand this comment. The term is positive by default hence is_negative = false. Negative signs makes negative numbers positive and positive numbers negative hence the is_negative = not is_negative.

VisusAdAstra replied:
ah i get it thanks

VisusAdAstra commented on parse.py:
```diff
@@ -0,0 +1,222 @@
+        # Expression grammars. Order of operations is determined here.
+        factor   = Forward()
+        atom     = (addition_operation[...] + ((function | value).setParseAction(self.push_first) | Group(left_bracket + expr + right_bracket))).setParseAction(self.push_unary_operator)
+        factor   << atom    + ((power_operation + factor) | factorial_operation).setParseAction(self.push_first)[...]
```
something wrong here for ex 2^5^2 and 2^3!^2 give wrong result

VisusAdAstra replied:
i will fix this one, mark solved for now

VisusAdAstra replied:
wrong understanding, its good

EmanuelUnity3D replied:
It works for me. I'm assuming that you might have assumed that exponents are read from left to right.

VisusAdAstra commented on parse.py:
```diff
@@ -0,0 +1,222 @@
+        # Expression grammars. Order of operations is determined here.
+        factor   = Forward()
+        atom     = (addition_operation[...] + ((function | value).setParseAction(self.push_first) | Group(left_bracket + expr + right_bracket))).setParseAction(self.push_unary_operator)
+        factor   << atom    + ((power_operation + factor) | factorial_operation).setParseAction(self.push_first)[...]
```
Based on this the exponent and factorial have the same precedence however when testing the factorial gets higher precedence? Can you explain for me why it happen?

EmanuelUnity3D replied:
I'm not sure why this would mean that exponents and factorials have the same precedence. The exponent of a power has to be a factor according to the definition above. The argument to a factorial can be an atom. Since factors must be parsed from atoms, this means that factorials will have higher precedence than powers.

VisusAdAstra replied:
here: atom    + ((power_operation + factor) | factorial_operation). The atom can be either of these form atom^factor or atom!. Therefore, from my pov they have the same precedence .

EmanuelUnity3D replied:
No because the rule states that exponents must be created from a factor power and that factorials only require an atom as an argument. Ergo, if you have atom^atom!, the factorial MUST be processed first because you cannot have a factor as the exponent in a power. Then you would have atom^factor at which point the power would be processed as well.

EmanuelUnity3D commented:
> plz add test suite to ensure proper functionality

@VisusAdAstra I agree that we need to do this but it's listed under a separate issue right now. Can we merge this for now and add that later?

SwatiPareek commented on parse.py:
```diff
@@ -0,0 +1,233 @@
+            operand_1 = self.evaluate_stack(symbol_stack)
+            return operation(operand_1, operand_2)
+        # Process function calls
+        operation = self.function_map.get(symbol, False)
```
Just curious, why did we switch from the original syntax of ... self.fn[op](*args)? 
Could just be me but that felt more readable.

EmanuelUnity3D replied:
This syntax allows us to account for operations not included in the operator dictionary.
